### PR TITLE
Start of PPPoE discovery code.

### PR DESCRIPTION
**Pull request by danderson**

### CODE REVIEW

File: ppp/discovery.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	}
+	ethernetBroadcast = &raw.Addr{
+		HardwareAddr: net.HardwareAddr{0xff, 0xff, 0xff, 0xff, 0xff, 0xff},
+	}
+)
+
+func pppoeDiscovery(ifName string) (sessionID uint16, err error) {
+	intf, err := net.InterfaceByName(ifName)
+	if err != nil {
+		return 0, fmt.Errorf("getting interface: %s", err)
```
Comment (bradfitz) [inline ppp/discovery.go]: conventionally we prefer %v for errors over %s

File: ppp/discovery.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
+package ppp
```
Comment (bradfitz) [inline ppp/discovery.go:2]: Missing package doc

File: ppp/discovery.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/mdlayher/raw"
+)
+
+const (
```
Comment (bradfitz) [inline ppp/discovery.go:12]: // Constants for XXXXX.

File: ppp/discovery.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	protoPPPoEDiscovery = 0x8863
+	protoPPPoESession   = 0x8864
+)
+
+var (
+	padiPacket = []byte{
```
Comment (bradfitz) [inline ppp/discovery.go:21]: // padiPacket is the PPPoE Active Discovery Initiation (PADI) packet.

File: ppp/discovery.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+		0x11, // Protocol version 1, packet type 1 (both constants)
+		9,    // Code PADI, aka "hey any PPPoE concentrators out there?"
+		0, 0, // Session ID not set during discovery
```
Comment (bradfitz) [inline ppp/discovery.go]: Session ID, not set ...

File: ppp/discovery.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+		return 0, fmt.Errorf("getting interface: %s", err)
+	}
+	conn, err := raw.ListenPacket(intf, protoPPPoEDiscovery, &raw.Config{LinuxSockDGRAM: true})
+	if err != nil {
+		return 0, fmt.Errorf("creating PPPoE Discovery listener: %s", err)
+	}
+	defer conn.Close()
+
+	if _, err := conn.WriteTo(padiPacket, ethernetBroadcast); err != nil {
+		return 0, fmt.Errorf("sending PADI packet: %s", err)
```
Comment (bradfitz) [inline ppp/discovery.go]: %v

File: ppp/discovery.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+	}
+
+	return 0, nil
```
Comment (bradfitz) [inline ppp/discovery.go]: panic("TODO") ?  I doubt 0 is the real sessionid.

File: ppp/ppp_test.go [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
+
+	return closeFunc, nil
+}
+
+func TestDiscovery(t *testing.T) {
+	if err := canTest(); err != nil {
+		t.Skipf("can't run privileged tests: %s", err)
```
Comment (bradfitz) [inline ppp/ppp_test.go]: %v

File: ppp/ppp_test.go [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+	}
+
+	close, err := startServer()
+	if err != nil {
+		t.Fatalf("couldn't start pppd container: %s", err)
+	}
+	defer close()
+	time.Sleep(200 * time.Millisecond)
```
Comment (bradfitz) [inline ppp/ppp_test.go]: Why?

**Review by bradfitz — DISMISSED** [review]

File: ppp/ppp_test.go (hunk 9 above)
Comment (danderson) [reply ppp/ppp_test.go]: I was debugging docker being weird. Eventually implemented the "wait for docker0 to come up" loop above and forgot to remove. Fixed.

**Comment by danderson** [comment]
Done.

File: ppp/ppp_test.go [hunk 10]
```diff
@@ -1,3 +1,4 @@
+// package ppp implements establishing a PPP connection running over PPPoE.
```
Comment (bradfitz) [inline ppp/ppp_test.go:1]: Capital Package. And s/implements establishing/establishes/.

**Comment by bradfitz** [comment]
But LGTM otherwise.

File: ppp/ppp_test.go (hunk 10 above)
Comment (danderson) [reply ppp/ppp_test.go:1]: Done.

**Pull request merged**